xdsl/dialects/builtin.py

```diff
@@ -2102,43 +2115,52 @@ def get_values(self) -> Sequence[int] | Sequence[float]:
         """
         Return all the values of the elements in this DenseIntOrFPElementsAttr
         """
-        return tuple(el.value.data for el in self.data.data)
+        return self.type.element_type.unpack(self.data.data, len(self))
 
     def get_attrs(self) -> Sequence[AnyIntegerAttr] | Sequence[AnyFloatAttr]:
         """
         Return all elements of the dense attribute in their relevant
         attribute representation (IntegerAttr / FloatAttr)
         """
-        return self.data.data
+        if isinstance(self.type.element_type, AnyFloat):
+            return tuple(
+                FloatAttr(value, self.type.element_type)
+                for value in cast(tuple[float, ...], self.get_values())
+            )
+        elif isinstance(self.type.element_type, IntegerType):
+            return tuple(
+                IntegerAttr(value, self.type.element_type)
+                for value in cast(tuple[int, ...], self.get_values())
+            )
+        else:  # index
+            return tuple(
+                IntegerAttr(value, self.type.element_type)
+                for value in cast(tuple[int, ...], self.get_values())
+            )
```

Comment:
This would be a great helper in a separate PR, I think, and also provided for DenseArrayBase. I wonder if we should change `iter/unpack` to `iter_values/unpack_values` and `iter_attrs/unpack_attrs`